File: 05/solution.ts
const input = await Bun.file(`${import.meta.dir}/input.txt`).text();
const [rangeStr, dataStr] = input
	.trim()
	.split("\n\n")
	.map((x) => x.split("\n"));

function inRange(r: number[], d: number) {
	const [min, max] = r;
	return d >= min && d <= max;
}

function part1() {
	const ranges = rangeStr.map((r) => r.split("-").map(Number));
	const data = dataStr.map(Number);
	let count = 0;

	for (const d of data) {
		if (ranges.some((r) => inRange(r, d))) {
			count++;
		}
	}

	return count;
}

function reduceRanges(ranges: number[][]) {
	// any pair of ranges whose end of one and start of the other is one apart can be replaced with a single range
	const sorted = [...ranges].sort((a, b) => a[0] - b[0]);
	const result: number[][] = [];

	for (const [start, end] of sorted) {
		const last = result.at(-1);
		if (last && start <= last[1] + 1) {
			last[1] = Math.max(last[1], end);
		} else {
			result.push([start, end]);
		}
	}

	return result;
}

// do a pass where I simplify the overlapping ranges so there are none overlapping
// then simply do a sum of differences of each range
function part2() {
	const ranges = rangeStr.map((r) => r.split("-").map(Number));
	const nonOverlapping: number[][] = reduceRanges(ranges);
	let sum = 0;

	for (const [start, end] of nonOverlapping) {
		sum += end - start + 1;
	}

	return sum;
}

console.log("Part 1:", part1());
console.log("Part 2:", part2());

File: 05/solution.ml
open Core
open Stdio

let read_input () =
  In_channel.read_lines "05/input.txt"
  |> List.split_while ~f:(fun s -> not (String.is_empty s))
  |> fun (a, b) -> (a, List.tl_exn b)

let parse_range s =
  match String.lsplit2 s ~on:'-' with
  | Some (a, b) -> (Int.of_string a, Int.of_string b)
  | None -> failwith "bad range"

let in_range (lo, hi) d = d >= lo && d <= hi

let reduce_ranges ranges =
  ranges
  |> List.sort ~compare:(fun (a, _) (b, _) -> Int.compare a b)
  |> List.fold ~init:[] ~f:(fun acc (start, end_) ->
      match acc with
      | (lo, hi) :: rest when start <= hi + 1 -> (lo, Int.max hi end_) :: rest
      | _ -> (start, end_) :: acc)

let part1 sranges sdata =
  let ranges = List.map sranges ~f:parse_range in
  let data = List.map sdata ~f:Int.of_string in
  data |> List.count ~f:(fun d -> List.exists ranges ~f:(fun r -> in_range r d))

let part2 sranges _sdata =
  let ranges =
    sranges
    |> List.map ~f:(fun s ->
        String.lsplit2_exn s ~on:'-' |> Tuple2.map ~f:Int.of_string)
  in
  reduce_ranges ranges |> List.sum (module Int) ~f:(fun (lo, hi) -> hi - lo + 1)

let () =
  let sranges, sdata = read_input () in
  printf "Part 1: %d\n" (part1 sranges sdata);
  printf "Part 2: %d\n" (part2 sranges sdata)
